### PR TITLE
[FEAT] Update methods to accept `BlockingRule` objects as arguments

### PR DESCRIPTION
**Pull request by ThomasHepworth**
### Type of PR

- [x] FEAT

### Is your Pull Request linked to an existing Issue or Pull Request?
This PR is a follow up to https://github.com/moj-analytical-services/splink/pull/1370, which introduces a blocking rule library.

### Give a brief description for the solution you have provided
This PR:
* Allows `estimate_parameters_using_expectation_maximisation` to take a `BlockingRule` class as an argument (meaning we can pass our pre-built `brl.exact_match_rule` rules as arguments).
* Adds support for the `BlockingRule` class to **hopefully** all of our existing methods.
* Updates a number of tests to make use of the blocking_rule_library functions.

### CODE REVIEW

**Comment by RossKen** [comment]
@ThomasHepworth  I would be interested to see the result of the conversation on slack on rules naming conventions before allowing `BlockingRules` to be passed to `estimate_parameters_using_expectation_maximisation`.

If we decide to stop calling the em rules blocking rules (training rules, for example) I think it would make sense to have a separate `TrainingRules` class which can be passed into `estimate_parameters_using_expectation_maximisation` instead. Where `BlockingRules` and `TrainingRules` would pretty much have the same properties (as they do the same thing under the hood), but are named separately to aid user understanding.

**Comment by RossKen** [comment]
Based off Robin and Sam's comments on slack, please disregard ☝️

File: tests/test_blocking.py [hunk 1]
```diff
@@ -52,3 +52,36 @@ def assess_preceding_rules(settings_brs_index):
...
     assess_preceding_rules(2)
     assess_preceding_rules(3)
+
+
+@mark_with_dialects_excluding()
+def test_simple_end_to_end(test_helpers, dialect):
+    helper = test_helpers[dialect]
+    Linker = helper.Linker
+    brl = helper.brl
+    df = helper.load_frame_from_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
```
Comment (RossKen) [inline tests/test_blocking.py:62]: update with inbuilt dataset

File: tests/test_estimate_prob_two_rr_match.py [hunk 2]
```diff
@@ -48,6 +49,15 @@ def test_prob_rr_match_dedupe(test_helpers, dialect):
         deterministic_rules, recall=0.9
     )
 
+    # test with BlockingRule object
+    deterministic_rules = [
```
Comment (RossKen) [inline tests/test_estimate_prob_two_rr_match.py:53]: For clarity of blocking rules we do not want blocking rule library to be compatible with estimating lambda

File: tests/test_find_new_matches.py [hunk 3]
```diff
@@ -59,22 +65,28 @@ def test_tf_tables_init_works():
...
-    linker = DuckDBLinker(
-        df,
-        settings,
-    )
+@mark_with_dialects_excluding()
+def test_matches_work(test_helpers, dialect):
+    helper = test_helpers[dialect]
+    Linker = helper.Linker
+    brl = helper.brl
+    df = helper.load_frame_from_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
```
Comment (RossKen) [inline tests/test_find_new_matches.py:73]: convert to in-built dataset

File: tests/test_full_example_deterministic_link.py [hunk 4]
```diff
@@ -1,26 +1,26 @@
...
-@pytest.mark.parametrize(
-    ("Linker"),
-    [
-        pytest.param(DuckDBLinker, id="DuckDB Deterministic Link Test"),
-        pytest.param(SparkLinker, id="Spark Deterministic Link Test"),
-    ],
-)
-def test_deterministic_link_full_example(tmp_path, spark, Linker):
-    df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
+    df = helper.load_frame_from_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
```
Comment (RossKen) [inline tests/test_full_example_deterministic_link.py:12]: convert to in-built

**Review by RossKen — COMMENTED** [review]
Hey, this is mostly good to go from my end. Thanks for your work!

Couple of small things:
- rebase and update tests with in built datasets
- remove capability for brl to be passed into prob two records match function. I think for clarity it is important to keep these separate. Is there some sort of warning for users if they do it?

File: tests/test_find_new_matches.py (hunk 3 above)
Comment (ThomasHepworth) [reply tests/test_find_new_matches.py:73]: Do we definitely want to? The in-built dataset is downloaded, so will add to the total run time of the tests.

Comment (ThomasHepworth) [reply tests/test_find_new_matches.py:73]: Unless you mean something else

Comment (RossKen) [reply tests/test_find_new_matches.py:73]: Good point - I hadn't thought about the additional runtime. Feel free to leave for now and we an have a more general conversation about how we use data for testing when you are back from leave. The key thing is that everything passes, which they do

**Pull request closed without merging**